### PR TITLE
GEOMESA-888 Fixing lazy transforms of kryo features when schema has been updated

### PR DESCRIPTION
**Pull request by elahrvivaz**
Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/accumulo/data/AccumuloDataStoreTransformsTest.scala [hunk 1]
```diff
@@ -246,5 +246,73 @@ class AccumuloDataStoreTransformsTest extends Specification with TestWithMultipl
...
+        features.map(_.getID) must containTheSameElementsAs(Seq("f1", "f2"))
+        features.sortBy(_.getID).map(_.getAttribute("geom").toString) mustEqual Seq("POINT (51 50)", "POINT (52 50)")
+        features.sortBy(_.getID).map(_.getAttribute("dtg")).map(new DateTime(_).withZone(DateTimeZone.UTC).getHourOfDay) mustEqual Seq(1, 2)
+      }
+      "for old attributes with new features" >> {
+        val query = new Query(sftName, ECQL.toFilter("IN ('f1')"), Array("geom", "dtg"))
+        val features = SelfClosingIterator(ds.getFeatureSource(sftName).getFeatures(query).features).toList
+        features.map(_.getID) must containTheSameElementsAs(Seq("f1"))
+        features.head.getAttribute("geom").toString mustEqual "POINT (51 50)"
+        new DateTime(features.head.getAttribute("dtg")).withZone(DateTimeZone.UTC).getHourOfDay mustEqual 1
```
Comment (apc3n) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/accumulo/data/AccumuloDataStoreTransformsTest.scala]: you use this above in a different way so you could extract a function to do _.getAttribute => HourOfDay

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/KryoBufferSimpleFeature.scala [hunk 2]
```diff
@@ -42,8 +42,9 @@ class KryoBufferSimpleFeature(sft: SimpleFeatureType,
   private var userDataOffset: Int = -1
 
   private var binaryTransform: () => Array[Byte] = input.getBuffer
+  private var reserializeTransform: () => Array[Byte] = input.getBuffer
```
Comment (apc3n) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/KryoBufferSimpleFeature.scala]: why are these functions?

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/KryoBufferSimpleFeature.scala [hunk 3]
```diff
@@ -42,8 +42,9 @@ class KryoBufferSimpleFeature(sft: SimpleFeatureType,
...
 
-  def transform(): Array[Byte] = binaryTransform()
+  def transform(): Array[Byte] = if (offsets.contains(-1)) reserializeTransform() else binaryTransform()
```
Comment (apc3n) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/KryoBufferSimpleFeature.scala]: same here why a method?

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/KryoBufferSimpleFeature.scala (hunk 2 above)
Comment (elahrvivaz) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/KryoBufferSimpleFeature.scala]: the functions get overwritten in the setTransform method - so they need to be vars.
Execution flow is:
val feature = // get reusable feature instance
feature.setTransform // set transforms once for this execution cycle
while (looping) { feature.setBuffer(newBytes); val result = feature.transform() }

File: geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/KryoBufferSimpleFeature.scala (hunk 3 above)
Comment (elahrvivaz) [inline geomesa-features/geomesa-feature-kryo/src/main/scala/org/locationtech/geomesa/features/kryo/KryoBufferSimpleFeature.scala]: this method exposes the private vars above, without allowing them to be modified.

**Comment by jnh5y** [comment]
Committed as 36604e893efc500ab67aa8748aba5ed827e0017e on master.

**Pull request closed without merging**